### PR TITLE
NH-15963-Provide-MongoDB-Instrumentation

### PR DESCRIPTION
**Pull request by ronilan**
## Overview

This pull request redoes and updates MongoDB instrumentation. Probe now instruments version 2.6 to 5 of mongoDB via two versons of the [`mongodb`](https://www.npmjs.com/package/mongodb) package.

It adds sanitization of mongodb queries.

## Status

* Support for auto instrumentation of MongoDB was added in [version 2.4 of the agent, back in 2015](https://github.com/appoptics/appoptics-apm-node/commit/ecba9bd1c5ff1c68cfc82bd52dfef2d4ea4a96b4) . This was done via [`mongodb-core`](https://npmjs.org/package/mongodb-core)    

* Four years later, in 2019, changes to driver implementation required creating instrumentation via the [`mongodb`](https://www.npmjs.com/package/mongodb) package .  

* Implementation was added to in [Pull Request 86](https://github.com/appoptics/appoptics-apm-node/pull/86).  

* There were issues with the implementation from the get go, and it has been known since early 2020 that the though the tests pass, instrumentation does not work in a variety of real world cases.

### mongodb probe

* Rewritten probe provide auto instrumentation of MongoDB version 2.6 to 5, using the two supported versions of the driver: [legacy 3.7.3](https://www.npmjs.com/package/mongodb/v/3.7.3) and [latest 4.7.0](https://www.npmjs.com/package/mongodb/v/4.8.1). See [MongoDB Node.js driver comparability chart](https://www.mongodb.com/docs/drivers/node/current/compatibility/). 

### Setup
  * Added Versions 4 and 5 of the mongodb database to dev environment and GitHub Actions setup.
  * Removed version 2.4.
  * Removed mongodb-core probe, tests and dependency.
  * Updated default mongodb package to latest (4.8.1). This is the one tests run against.

### Features
  * Added recursive object sanitizer that replace strings, string buffers and numbers with a `?`.
  * A probe configuration value: `sanitizeObject` was added to mongodb probe config options. It is `true` by default.

## Notes

- MongoDB version 6 was released [a couple of weeks back](https://www.mongodb.com/blog/post/mongodb-6-0-now-available). It is not included in this pull request as the official docker image for `mongo:latest` is still at 5.

### CODE REVIEW

File: test/object-sanitizer.test.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  })
+
+  it('should sanitizes strings', function () {
+    const obj = { a: 'hide', b: 'me' }
+    const result = objectSanitizer.sanitize(obj)
+    expect(JSON.stringify(result)).equal(JSON.stringify({ a: '?', b: '?' }))
+  })
+
+  it('should sanitizes double quoted strings', function () {
+    const obj = { a: 'hide', b: 'me' }
```
Comment (tammy-baylis-swi) [inline test/object-sanitizer.test.js]: I think `obj` should be `{ a: "hide", b: "me" }` here, else I think this test is the same as the previous one.

File: test/probes/mongoose.test.js [hunk 2]
```diff
@@ -295,7 +295,7 @@ describe(`probes/mongoose ${pkg.version} using ${moduleName}`, function () {
     // 2. issue multiple inserts that must be queued because the connection is pending
     // 3. wait for all inserts to complete before issuing the done.
     //
-    it(`should connect and queue queries using a ${type}`, function (realDone) {
+    it.skip(`should connect and queue queries using a ${type}`, function (realDone) {
```
Comment (tammy-baylis-swi) [inline test/probes/mongoose.test.js]: Does this skip the test, and should we skip this test?

File: lib/probes/mongodb.js [hunk 3]
```diff
@@ -1,347 +1,340 @@
...
+      QueryOp: 'remove',
+      Query: prepQueryString(obj.map(v => v.q))
     }
   }
+}
 
-  data.QueryOp = 'command'
-  data.Command = JSON.stringify(cmd)
+module.exports = function (mongodb, options) {
+  // v3.3.0 is the first version of mongodb that doesn't use mongodb-core.
```
Comment (tammy-baylis-swi) [inline lib/probes/mongodb.js:204]: _Basic_ questions from me. Not sure where this is in the code and/or compatibility chart.

The agent supports Mongo `2.6` and `3.0`, `3.1`, `3.2` as well, which I think do use `mongodb-core` is that right? Generally, what happens in the agent if customer is using one of those older versions with this PR update? e.g. Are all commands and values sanitized the same way as `>=3.3.0`? How come `requires('mongodb-core')` is no longer needed?

File: test/probes/mongoose.test.js (hunk 2 above)
Comment (ronilan) [reply test/probes/mongoose.test.js]: It skips the test. 

Mongoose is an ODM (i.e. ORM of no-sql dbs) that also has issues that needs rework. The test that was failing while WIP,. But it is good now... will unskip.

**Comment by tammy-baylis-swi** [comment]
Overall lgtm; posted one question set and two suggestions for the tests.

Based on what I think this is doing it generally makes sense to me. No more `<2.6` support as you wrote in PR description and the Docker envs are doing the same. Supporting the differently-named commands of Mongo `3` vs `4` is a feat!

File: test/object-sanitizer.test.js (hunk 1 above)
Comment (ronilan) [reply test/object-sanitizer.test.js]: The linter fixed it.. :D
Will prevent all no linting comment.

File: lib/probes/mongodb.js (hunk 3 above)
Comment (ronilan) [reply lib/probes/mongodb.js:204]: mongodb the *database* versions 2.6 to 5
mongodb the *driver* https://www.npmjs.com/package/mongodb 3.3.0 and above.

Customer who has older version of *the database* should *upgrade the driver* using `npm install mongodob@legacy` (which will give them driver `3.7.3`). Then from instrumentation point of view the 'mongodb-core' package is dropped. In short - we do not support outdated legacy configurations.

Comment (tammy-baylis-swi) [reply lib/probes/mongodb.js:204]: Hah, key piece I was missing. Thank you!

**Review by tammy-baylis-swi — APPROVED** [review]
Thanks for the test tweaks!

**Pull request merged**